engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/UpdateBy.java

```diff
@@ -332,10 +335,21 @@ class PhasedUpdateProcessor implements LogOutputAppendable {
             } else {
                 // Determine which windows need to be computed.
                 for (int winIdx = 0; winIdx < windows.length; winIdx++) {
-                    final int fWinIdx = winIdx;
-                    // look in the dirty buckets for the windows that need to be computed
-                    dirtyWindows[fWinIdx] = Arrays.stream(dirtyBuckets)
-                            .anyMatch(bucket -> windows[fWinIdx].isWindowBucketDirty(bucket.windowContexts[fWinIdx]));
+                    final int finalWinIdx = winIdx;
+                    if (Arrays.stream(dirtyBuckets)
+                            .anyMatch(bucket -> windows[finalWinIdx]
+                                    .isWindowBucketDirty(bucket.windowContexts[finalWinIdx]))) {
+                        dirtyWindows.set(winIdx);
+                        dirtyWindowOperators[winIdx] = new BitSet(windows[winIdx].operators.length);
+                        final int size = windows[winIdx].operators.length;
+                        for (UpdateByBucketHelper bucket : dirtyBuckets) {
+                            dirtyWindowOperators[winIdx].or(bucket.windowContexts[winIdx].dirtyOperators);
+                            if (dirtyWindowOperators[winIdx].cardinality() == size) {
+                                // all are set, we can stop checking
+                                break;
+                            }
+                        }
+                    }
```

Comment:
Maybe we could iterate the dirty buckets only once?
```suggestion
                for (int winIdx = 0; winIdx < windows.length; winIdx++) {
                    for (UpdateByBucketHelper bucket : dirtyBuckets) {
                        final UpdateByWindow.UpdateByWindowBucketContext bucketWindowCtx = bucket.windowContexts[winIdx];
                        if (!bucketWindowCtx.isDirty) {
                            continue;
                        }
                        if (dirtyWindowOperators[winIdx] == null) {
                            dirtyWindows.set(winIdx);
                            dirtyWindowOperators[winIdx] = new BitSet(windows[winIdx].operators.length);
                        }
                        final int size = windows[winIdx].operators.length;
                        dirtyWindowOperators[winIdx].or(bucketWindowCtx.dirtyOperators);
                        if (dirtyWindowOperators[winIdx].cardinality() == size) {
                            // all are set, we can stop checking
                            break;
                        }
                    }
                }
```